drivers: gpio: add "gpio_reserved_ranges" property support

avolkov-1221 opened the pull request:
Added missing support of DT gpio's "gpio_reserved_ranges" property. The array of pin range pairs as shown below now works. The implementation constraint is inherited from common DT stuff: a maximum of 64 pairs can be used, i.e. no more than 128-bit gpio ports are supported.

Usage example:
The reserved mask , defined in the device tree as:
gpio0: gpio0 {
....
ngpios=<32>;
gpio-reserved-ranges = <0 8>, <9 5>, <15 17>;
....
};

and used in driver configuration as:
{
...
.gpio_reserved_ranges = GPIO_RESERVED_PIN_MASK_FROM_DT_NODE(
DT_NODELABEL(gpio0)),
...
}

correctly converts to 0x80004100 (0b10000000000000000100000100000000)

Fixes #56999

henrikbrixandersen commented:
@avolkov-1221 Thank you for your proposal. You only submitted it yesterday, but we will get to it. 

Please see our documentation for more details on what to expect time wise when submitting a PR: https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-review-escalation

avolkov-1221 commented:
> @avolkov-1221 Thank you for your proposal. You only submitted it yesterday, but we will get to it.
> 
> Please see our documentation for more details on what to expect time wise when submitting a PR: https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-review-escalation

Ok, thank you, I'll be less impatient from now on :).

avolkov-1221 commented on include/zephyr/drivers/gpio/gpio_utils.h:
```diff
@@ -26,6 +26,43 @@
+				GET_ARG_N(idx, Z_LIST_EVEN_ARGS), idx)), \
+		(0))
+
+#define GPIO_RESERVED_PIN_MASK_FROM_DT_NODE(node_id)			      \
+	COND_CODE_1(DT_NODE_HAS_PROP(node_id, gpio_reserved_ranges),	      \
+		((GPIO_PORT_PIN_MASK_FROM_DT_NODE(node_id) &		      \
+			~(FOR_EACH_FIXED_ARG(GEN_GPIO_RESERVED_PIN_MASK_COND, \
+					(|),				      \
+					node_id,			      \
+					Z_LIST_ODD_NUMBERS)))),		      \
+		(GPIO_PORT_PIN_MASK_FROM_DT_NODE(node_id)))
+
+#define GPIO_RESERVED_PIN_MASK_FROM_DT_INST(inst)			\
+		GPIO_RESERVED_PIN_MASK_FROM_DT_NODE(DT_DRV_INST(inst))
```
Hmm, for the test plan is unclear what exactly I can do, except simple 0, 0xff and 0x55 patterns.
Most interesting cases: bitfields overlaps/out of range and odd number values, looks too hard to implement without some error reporting, but something like #error is useless in DT, isn't it?

Doxygen  comments I'll add.

avolkov-1221 replied:
PS: BTW I rethought and want to rename GPIO_RESERVED_PIN_MASK_xxx to GPIO_RESERVED_RANGES_xxx, to follow the property name

talih0 replied:
> 

Maybe something like this
GPIO_PORT_PIN_MASK_EXC_RESERVED_RANGES_xxx

avolkov-1221 replied:
> > 
> 
> Maybe something like this? GPIO_PORT_PIN_MASK_EXC_RESERVED_RANGES_xxx

Hmm, I would disagree: this name tells nothing about what this macro exactly does, unlike my last one: it returns the value of the property in the form of a bitmask, according to the property's description. 

BTW I've changed not only the name, but also the behavior: now the macro returns exactly the bitmask as it is defined in DT. Also I added two properties with above behavior, specifically for defining ports with forbidden pins (for use, as an example, in .port_pin_mask from gpio_driver_config definition). They are called DT_GPIO_PORT_PIN_MASK and DT_INST_GPIO_PORT_PIN_MASK respectively.

An updated commit will be sent tomorrow, I need to finish the tests.

talih0 commented:
nice, work!

I think we should change the meaning of the macro:
#define GPIO_PORT_PIN_MASK_FROM_DT_NODE() to your implementation that excludes the reserved ranges.

Then my PR (https://github.com/zephyrproject-rtos/zephyr/pull/55435), wouldn't have to do any changes once your PR gets merged.

gpio_nxp_s32.c and gpio_mcux_igpio.c should benefit from your compile time implementation.

talih0 commented:
> and used in driver configuration as: { ... .gpio_reserved_ranges = GPIO_RESERVED_PIN_MASK_FROM_DT_NODE( DT_NODELABEL(gpio0)), ... }
> 
> correctly converts to 0x80004100 (0b10000000000000000100000100000000)

Shouldn't this be 0x4100?

avolkov-1221 commented:
> > and used in driver configuration as: { ... .gpio_reserved_ranges = GPIO_RESERVED_PIN_MASK_FROM_DT_NODE( DT_NODELABEL(gpio0)), ... }
> > correctly converts to 0x80004100 (0b10000000000000000100000100000000)
> 
> Shouldn't this be 0x4100?

Ah yes, thank you, it's a copy paste bug, actually my commit's example should be .gpio-reserved-ranges = <0 8>, <9 5>, <15 1**6**>;

cfriedt commented:
Looks reasonable enough to me - there is some minor coding style issue though.

avolkov-1221 commented:
> Looks reasonable enough to me - there is some minor coding style issue though.

I hope this is fixed in the last patch series. 

IMHO it was 'false positive' of checkpatch.pl

avolkov-1221 commented:
@henrikbrixandersen Hello Brix,
Did you have any objections vs merge? 
Or should I wait one week more before raising the PR status?

talih0 commented:
@avolkov-1221, thank you for spending time to document the macros.

I still have the same comment: I think the default behavior of
GPIO_PORT_PIN_MASK_FROM_DT_INST() should be changed so that the reserved ranges are removed by default. At the end of the day, gpio drivers just care about properly computing the pin_mask so that the driver can error out if an invalid pin is used.

Also, I'm not sure if gpio.h is the right place for the macros. Wouldn't  gpio_utils.h be a better place where the current mask macros are defined? (as you had in your origin pr)

mbolivar-nordic commented on include/zephyr/devicetree/gpio.h:
```diff
@@ -438,6 +438,198 @@ extern "C" {
+ * @param node_id GPIO controller node identifier.
+ * @return the bitmask of reserved gpios
+ */
+#define DT_GPIO_RESERVED_RANGES(node_id)	((gpio_port_pins_t)	       \
```
Hmm, no, I'm sorry to jump in late here but this is not acceptable to go into the devicetree API. I did not receive a review request until now for some reason.

The dependency layering is that gpio.h can depend on devicetree.h, but not the other way around. Please rework the PR so that anything which need to know anything about the Zephyr GPIO API is not in devicetree.h or any of its sub-headers like this one.

avolkov-1221 replied:
> Hmm, no, I'm sorry to jump in late here but this is not acceptable to go into the devicetree API. I did not receive a review request until recently for some reason.

You are welcome, however somehow I have a feeling that my pebble caused an avalanche... :)

> 
> The dependency layering is that gpio.h can depend on devicetree.h, but not the other way around. Please rework the PR so that anything which need to know anything about the Zephyr GPIO API is not in devicetree.h or any of its sub-headers like this one.

Ah, I got it, but then where would it be better to put this things: into the **drivers/gpio.h** or return it back into **drivers/gpio/gpio_utils.h** ?

@talih0 IMO **drivers/gpio.h** is better suited for such type of thing as it's a missing part of the general GPIO API and not something ancillary as expected from the code placed in the **xxxx_utils.h** headers. However I don't know the subject's prehistory so suggestions are welcome.

@mbolivar-nordic should I also change the macro prefix from DT_GPIO_xx to GPIO_DT_xx, as is common in "drivers/gpio.h"? BTW it looks a bit confusing that we have two unrelated sets (DT_GPIO_ and GPIO_DT_) to describe parts of a whole. 
And as a consequence, is it acceptable to keep my tests in **tests/lib/devicetree/api**, or they have to relocated into the **tests/drivers/gpio/** ?

mbolivar-nordic commented on include/zephyr/devicetree/gpio.h:
```diff
@@ -438,6 +438,198 @@ extern "C" {
+ * @param node_id GPIO controller node identifier.
+ * @return the bitmask of allowed gpios
+ */
+#define DT_GPIO_PORT_PIN_MASK_EXC(node_id)	((gpio_port_pins_t)	       \
```
Same here; this is not acceptable for the DT API

avolkov-1221 commented:
> @avolkov-1221, thank you for spending time to document the macros.
> 
> I still have the same comment: I think the default behavior of GPIO_PORT_PIN_MASK_FROM_DT_INST() should be changed so that the reserved ranges are removed by default. At the end of the day, gpio drivers just care about properly computing the pin_mask so that the driver can error out if an invalid pin is used.
> 
I agree and also I think that GPIO_PORT_PIN_MASK_FROM_DT_INST is useless and obsoleted now. But as well I'm not sure that this simple modification will not bring some side effects in some module. The problem is further aggravated by the fact that there is not even a single formal test for these macros, I suspect because they were added too long ago.
@galak @mbolivar-nordic @carlescufi Could you please comment on this?

So probably it will be better to mark GPIO_PORT_PIN_MASK_FROM_xxxx macro as deprecated, and replace it everywhere by the new GPIO_DT_PORT_PIN_MASK_EXC (or DT_GPIO_PORT_PIN_MASK_EXC). The latter, I suspect, will cause a very large and long discussion/review (almost all files in **drivers/gpio** are affected), so IMHO it would be better to make it as a separate PR, with integration, probably, in the next release. Not sure if a month will be  enough (3.4 is planned for June).

talih0 commented:
> 



> > @avolkov-1221, thank you for spending time to document the macros.
> > I still have the same comment: I think the default behavior of GPIO_PORT_PIN_MASK_FROM_DT_INST() should be changed so that the reserved ranges are removed by default. At the end of the day, gpio drivers just care about properly computing the pin_mask so that the driver can error out if an invalid pin is used.
> 
> I agree and also I think that GPIO_PORT_PIN_MASK_FROM_DT_INST is useless and obsoleted now. But as well I'm not sure that this simple modification will not bring some side effects in some module. The problem is further aggravated by the fact that there is not even a single formal test for these macros, I suspect because they were added too long ago. @galak @mbolivar-nordic @carlescufi Could you please comment on this?
> 

It should be safe because the output is equivalent if the user does not specify a reserved range in the device tree.

mbolivar-nordic commented:
On Wed, Apr 12 2023, Andrey VOLKOV wrote:
> @avolkov-1221 commented on this pull request.
>
>
>
>> + *		.ngpios = 32,
> + *		.gpios_reserved = 0xdeadbeef,
> + *		               // 0b1101 1110 1010 1101 1011 1110 1110 1111
> + *
> + *	static const struct some_config dev_cfg_b = {
> + *		.ngpios = 18,
> + *		.gpios_reserved = 0xfffc0418,
> + *		               // 0b1111 1111 1111 1100 0000 0100 0001 1000
> + *		               // unused MSBs were marked as reserved too
> + *	};
> + * @endcode
> + *
> + * @param node_id GPIO controller node identifier.
> + * @return the bitmask of reserved gpios
> + */
> +#define DT_GPIO_RESERVED_RANGES(node_id)	((gpio_port_pins_t)	       \
>
>> Hmm, no, I'm sorry to jump in late here but this is not acceptable to go into the devicetree API. I did not receive a review request until recently for some reason.
>
> You are welcome, however somehow I have a feeling that my pebble
> caused an avalanche... :)

It is often my experience in Zephyr that if you pull on a tiny little
thread, you'll end up with a rope in your hands...

>
>>
>> The dependency layering is that gpio.h can depend on devicetree.h, but not the other way around. Please rework the PR so that anything which need to know anything about the Zephyr GPIO API is not in devicetree.h or any of its sub-headers like this one.
>
> Ah, I got it, but then where would it be better to put this things:
> into the **drivers/gpio.h** or return it back into
> **drivers/gpio/gpio_utils.h** ?

I am only a devicetree maintainer, so the GPIO API is not my area of
responsibility. But FWIW I think the decision you have made in the
latest push makes good sense.

> @mbolivar-nordic should I also change the macro prefix from DT_GPIO_xx
> to GPIO_DT_xx, as is common in "drivers/gpio.h"? BTW it looks a bit
> confusing that we have two unrelated sets (DT_GPIO_ and GPIO_DT_) to
> describe parts of a whole.

This is intentional: the DT_ namespace is reserved for
<zephyr/devicetree.h>. So DT_GPIO_ is the namespace of macros in the
devicetree API that are somehow related to GPIOs. Defining a macro that
begins with DT_ outside of this header and its sub-headers is improper
use of the DT namespace.

The GPIO_ namespace belongs to <zephyr/drivers/gpio.h>. So GPIO_DT_ is a
namespace for GPIO APIs that are somehow related to devicetree.

The latest push uses the namespaces correctly.

> And as a consequence, is it acceptable to keep my tests in
> **tests/lib/devicetree/api**, or they have to relocated into the
> **tests/drivers/gpio/** ?


In the latest push, nothing is touching the devicetree API, so I agree
that tests/drivers/gpio is the correct location.

I suggest changing the title of the last commit from:

```
tests: devicetree: ...
```

to:

```
tests: gpio: ...
```

accordingly, but this is is not a blocking request.

mbolivar-nordic commented:
@henrikbrixandersen please revisit

henrikbrixandersen commented on include/zephyr/drivers/gpio.h:
```diff
@@ -417,6 +417,252 @@ struct gpio_dt_spec {
+				DT_PROP_BY_IDX(node_id, prop, idx))))	     \
+		), (0))
+
+/*
+ * A helper conditional macro returning generated bitmask for
+ * one element from @p "gpio-reserved-ranges"
+ */
+#define GEN_GPIO_RESERVED_RANGES_COND(idx, node_id)			   \
```
Please document the parameters used.

avolkov-1221 replied:
done.

henrikbrixandersen commented on include/zephyr/drivers/gpio.h:
```diff
@@ -417,6 +417,252 @@ struct gpio_dt_spec {
+ * @param node_id GPIO controller node identifier.
+ * @return the bitmask of allowed gpios
+ */
+#define GPIO_DT_PORT_PIN_MASK_EXC(node_id)				       \
```
Why not roll this into the existing `GPIO_PORT_PIN_MASK_FROM_NGPIOS()`? It would seem with the current approach, each driver must be modified in order to support reserved GPIOs?

avolkov-1221 replied:
I'm supposing to do this as a separate PR: exactly because all gpios drivers will be affected (and some unknown external modules too), so it can be revert back in case of a regression. The current PR does not affect anything except our code for now.

avolkov-1221 replied:
I mean this one:
` #define GPIO_PORT_PIN_MASK_FROM_NGPIOS()           GPIO_DT_PORT_PIN_MASK_NGPIOS_EXC()`

talih0 replied:
> I'm supposing to do this as a separate PR: exactly because all gpios drivers will be affected (and some unknown external modules too), so it can be revert back in case of a regression. The current PR does not affect anything except our code for now.

why a new PR? maybe just make an extra commit

avolkov-1221 replied:
Okay, agreed, I'll do it.

avolkov-1221 commented on include/zephyr/drivers/gpio.h:
```diff
@@ -417,6 +417,252 @@ struct gpio_dt_spec {
+ * of @p "gpio-reserved-ranges" defined by the index pair, or 0 if the
+ * range does not exist.
+ */
+#define GEN_PIN_MASK_COND(node_id, prop, idx, idx_next)			     \
```
done.

avolkov-1221 commented on include/zephyr/drivers/gpio.h:
```diff
@@ -417,6 +417,252 @@ struct gpio_dt_spec {
+ * For example, setting "gpio-reserved-ranges = <3 2>, <10 1>;"
+ * means that GPIO offsets 3, 4 and 10 cannot be used even if @p ngpios = <18>.
+ *
+ * The implementation constraint is inherited from common DT stuff: a maximum
```
Fixed.

talih0 commented on include/zephyr/drivers/gpio/gpio_utils.h:
```diff
@@ -17,14 +17,16 @@
 
 #define GPIO_PORT_PIN_MASK_FROM_DT_NODE(node_id)		\
-	GPIO_PORT_PIN_MASK_FROM_NGPIOS(DT_PROP(node_id, ngpios))
+	GPIO_DT_PORT_PIN_MASK_EXC(node_id)
```
This change, and the change below are the only ones that need to be made in this commit.
I wouldn't even deprecate the macro GPIO_PORT_PIN_MASK_FROM_NGPIOS() or make any changes to the drivers.

avolkov-1221 replied:
I would not agree. GPIO_PORT_PIN_MASK_FROM_NGPIOS probably is not deprecated itself, but the way how it's using definitely is : 

* drivers/gpio/gpio_ite_it8xxx2.c and drivers/gpio/gpio_kscan_ite_it8xxx2.c IMHO were suddenly skipped by sed script from        4e553f31fa43efe0272f2e26935e2cbaf4f0c1fb commit,
* drivers/gpio/gpio_stm32.c and drivers/gpio/gpio_npcx.c are using the gpio_controller YAML, so should handle the `gpio-reserved-ranges` too
 *  only drivers/gpio/gpio xlnx axi.c has non-standard syntax, so @henrikbrixandersen should makes the decision.
 
Btw it's exactly why I resisted to make this modification as a commit, but a PR :). Let's flame go on...

talih0 replied:
> I would not agree. GPIO_PORT_PIN_MASK_FROM_NGPIOS probably is not deprecated itself, but the way how it's using definitely is :
> 
>     * drivers/gpio/gpio_ite_it8xxx2.c and drivers/gpio/gpio_kscan_ite_it8xxx2.c IMHO were suddenly skipped by sed script from        [4e553f3](https://github.com/zephyrproject-rtos/zephyr/commit/4e553f31fa43efe0272f2e26935e2cbaf4f0c1fb) commit,
> 
>     * drivers/gpio/gpio_stm32.c and drivers/gpio/gpio_npcx.c are using the gpio_controller YAML, so should handle the `gpio-reserved-ranges` too
> 
>     * only drivers/gpio/gpio xlnx axi.c has non-standard syntax, so @henrikbrixandersen should makes the decision.
> 
> 
> Btw it's exactly why I resisted to make this modification as a commit, but a PR :). Let's flame go on...

You can create an issue to track drivers that need to be updated.
I believe I made my suggested changes clear.

avolkov-1221 replied:
Hmm, sound reasonable for me. Ok. I'll revert deprecated mark.

avolkov-1221 replied:
Since, unfortunately, the standard Zephyr's create issue->create PR->merge procedure will delay the adoption of changes indefinitely, I'm returning the changes of the  drivers.
At least this PR has a chance of being accepted for the foreseeable future.

henrikbrixandersen commented on include/zephyr/drivers/gpio/gpio_utils.h:
```diff
@@ -21,10 +21,10 @@
 
 #define GPIO_PORT_PIN_MASK_FROM_DT_NODE(node_id)		\
-	GPIO_PORT_PIN_MASK_FROM_NGPIOS(DT_PROP(node_id, ngpios))
+	GPIO_DT_PORT_PIN_MASK_EXC(node_id)
 
 #define GPIO_PORT_PIN_MASK_FROM_DT_INST(inst)			\
-	GPIO_PORT_PIN_MASK_FROM_NGPIOS(DT_INST_PROP(inst, ngpios))
+	GPIO_DT_INST_PORT_PIN_MASK_EXC(inst)
```
Why are the aliases needed? Why not just change the implementation of these and leave out the `_EXC` macros?

avolkov-1221 replied:
The names "GPIO_PORT_PIN_MASK_FROM_NGPIOS" don't follow the GPIO API naming conventions at all, so IMHO it's  little bit confusing to use 3 different prefixes for the same API: DT_GPIO, GPIO_DT and GPIO_PORT. If for the first part, the reasons was clearly described by Marti Bolivar, the least one is an exception arrived from 2.1.xx epoch. So it's probably better to stop using them and switching the the new macros instead, which following to the common naming scheme and are in the expected location.

However, not touching existing code is also a reasonable decision, and that's exactly what aliasing does.

avolkov-1221 replied:
Sorry @galak @henrikbrixandersen for bothering you, but I don't have any response for another week.

avolkov-1221 replied:
Hello  @henrikbrixandersen, @nashif, @carlescufi, @galak, @MaureenHelm, @mnkp, @erwango,

Sorry to remind you but this PR has been on ready for over almost a month and on blocking without any response for over one week, is it still useful for Zephyr? And is one response per week a normal code review practice?

Regards,
Andrey VOLKOV

henrikbrixandersen reviewed (changes requested):
I have given this PR much thought, but my previous comments still stands:
- We already have `struct gpio_driver_config`->`port_pin_mask` for describing the mask of supported pins for a given GPIO controller driver instance.
- We already have macros for populating this mask from devicetree properties (`GPIO_PORT_PIN_MASK_FROM_DT_NODE()` and `GPIO_PORT_PIN_MASK_FROM_DT_INST()`.

For introducing generic support for reserved ranges for GPIO controller pins in DTS we need:
1. An optional `gpio-reserved-ranges` property in `dts/bindings/gpio/gpio-controller.yaml` (already added by @galak in 11966e599ed8c9d096e6754239917269b9b67e1c).
2. Changing the above two macros (`GPIO_PORT_PIN_MASK_FROM_DT_NODE()` and `GPIO_PORT_PIN_MASK_FROM_DT_INST()`) to factor in the `gpio-reverved-ranges` property when determining the pin mask. This means they can no longer just call `GPIO_PORT_PIN_MASK_FROM_NGPIOS()`, but likely need a new implementation. There is no need to introduce new top-level/user-facing macros in order to implement this as we already have the needed macros in place, they just need to take one more DTS property into considering for determining the pin mask. We should not have top-level macros as pure aliases for other top-level macros (unless they are deprecated, which is an entirely different discussion). Changing these two macros will allow us to introduce generic support for reserved GPIO pin ranges without touching (the majority of the) GPIO drivers.

cfriedt reviewed (changes requested):
I agree with @henrikbrixandersen 's take. I also feel that it might be smart to enforce an either / or use of the existing pin mask property or reserved ranges, but not both. That might be simpler than trying to skin the cat two different ways simultaneously.

carlescufi commented:
@avolkov-1221 I see that @henrikbrixandersen and @cfriedt have replied to you describing their concerns. If you'd like to bring this to the Architecture WG meeting to have it discussed there please let me know on Discord.

henrikbrixandersen commented on include/zephyr/drivers/gpio.h:
```diff
@@ -417,6 +417,240 @@ struct gpio_dt_spec {
+ * @param off_idx logical index of bitmask offset value into "prop" array
+ * @param sz_idx logical index of bitmask size value into "prop" array
+ */
+#define GEN_BITMASK_COND(node_id, prop, off_idx, sz_idx)		     \
```
Since this is a public header, please prefix the internal macros with `Z_NAMESPACE` (i.e. `Z_GPIO`) and exclude them from public API documentation using `@cond INTERNAL_HIDDEN`. This goes for all internal macros below. Only the user-facing macros should be included in the API documentation.

```suggestion
#define Z_GPIO_GEN_BITMASK_COND(node_id, prop, off_idx, sz_idx)		     \
```

avolkov-1221 replied:
Done

avolkov-1221 replied:
@henrikbrixandersen @cfriedt Sorry guys, but I'm waiting again for more then 2 week. @henrikbrixandersen Could you please finishing the review?

henrikbrixandersen replied:
Hi Andrey,

We are still waiting for you to implement the requested changes (use `Z_GPIO_` as prefix for internal macros, hide them from the API documentation using `@cond INTERNAL_HIDDEN`). Beware, we are currently in feature freeze for Zephyr v3.4.0, so people are focusing on fixing bugs prior to the release.

avolkov-1221 replied:
Hi @henrikbrixandersen ,
Sorry, I've forget @cond  part of your suggestion. It has been done now, so we can unlock the merge process.

henrikbrixandersen reviewed (changes requested):
Minor naming nit, otherwise looks good.

avolkov-1221 committed:
include: util_internal: add Z_SPARSE_LIST_{ODD|EVEN}_NUMBERS

Add auxiliary lists Z_SPARSE_LIST_ODD_NUMBERS and
Z_SPARSE_LIST_EVEN_NUMBERS. These lists were originally created for
GPIO_DT_RESERVED_RANGES_NGPIOS macro, but may be useful for others
applications too.

Signed-off-by: Andrey VOLKOV <andrey.volkov@munic.io>

avolkov-1221 committed:
drivers: gpio: add "gpio-reserved-ranges" property support

Added

   * GPIO_DT_RESERVED_RANGES_NGPIOS(node_id, ngpios),
   * GPIO_DT_RESERVED_RANGES(node_id),
   * GPIO_DT_INST_RESERVED_RANGES_NGPIOS(inst, ngpios) and
   * GPIO_DT_INST_RESERVED_RANGES(inst)

as DT gpio's "gpio-reserved-ranges" property missing support.
The array of pins range pairs as shown below now works.

Also implemented paired extended versions of
GPIO_PORT_PIN_MASK_FROM_NGPIOS macro:

   * GPIO_DT_PORT_PIN_MASK_NGPIOS_EXC(node_id, ngpios) and
   * GPIO_DT_INST_PORT_PIN_MASK_NGPIOS_EXC(inst, ngpios)

with reserved (ie EXCluded) pins support too.

The implementation constraint is inherited from common DT limitations:
a maximum of 64 pairs can be used, i.e. theoretically no more than
128-bit gpio ports are supported (but resulted bitmask is actually
limited by the size of gpio_port_pins_t type).

Usage example:
The reserved mask, defined in the device tree as:
    gpio0: gpio0 {
       ....
       ngpios=<32>;
       gpio-reserved-ranges = <0 8>, <9 5>, <15 16>;
       ....
    };

and used in driver configuration as:
    {
     ...
     .gpio_reserved = GPIO_DT_RESERVED_RANGES(DT_NODELABEL(gpio0)),
     .pin_port_mask = GPIO_DT_PORT_PIN_MASK_NGPIOS_EXC(
				DT_NODELABEL(gpio0),
				DT_PROP(DT_NODELABEL(gpio0), ngpios)),
     ...
    }

correctly converts to:

    .gpio_reserved = 0x7fffbeff, 0b01111111 11111111 10111110 11111111)
    .pin_port_mask = 0x80004100, 0b10000000 00000000 01000001 00000000)

Signed-off-by: Andrey VOLKOV <andrey.volkov@munic.io>

avolkov-1221 committed:
gpio: add "gpio-reserved-ranges" support to GPIO_PORT_PIN_MASK_XX

Updated the implementation of the GPIO_PORT_PIN_MASK_FROM_DT_NODE and
GPIO_PORT_PIN_MASK_FROM_DT_INST macros to support
"gpio-reserved-ranges" device tree property.

Signed-off-by: Andrey VOLKOV <andrey.volkov@munic.io>

avolkov-1221 committed:
tests: gpio: add tests for GPIO_DT_RESERVED_RANGES_NGPIOS et al

Added basic test coverage for the:

   * GPIO_DT_RESERVED_RANGES_NGPIOS,
   * GPIO_DT_RESERVED_RANGES,
   * GPIO_DT_INST_RESERVED_RANGES_NGPIOS,
   * GPIO_DT_INST_RESERVED_RANGES,
   * GPIO_DT_PORT_PIN_MASK_NGPIOS_EXC,
   * GPIO_PORT_PIN_MASK_FROM_DT_NODE,
   * GPIO_DT_INST_PORT_PIN_MASK_NGPIOS_EXC and
   * GPIO_PORT_PIN_MASK_FROM_DT_INST macros.

Signed-off-by: Andrey VOLKOV <andrey.volkov@munic.io>

henrikbrixandersen commented on include/zephyr/drivers/gpio.h:
```diff
@@ -417,6 +417,247 @@ struct gpio_dt_spec {
 #define GPIO_DT_SPEC_INST_GET_OR(inst, prop, default_value) \
 	GPIO_DT_SPEC_INST_GET_BY_IDX_OR(inst, prop, 0, default_value)
 
+/*
+ * @cond INTERNAL_HIDDEN
+ *
+ * Auxiliary conditional macro that generates a bitmask for the range
+ * from @p "prop" array defined by the (off_idx, sz_idx) pair,
+ * or 0 if the range does not exist.
+ *
+ * @param node_id devicetree node identifier
+ * @param prop lowercase-and-underscores array property name
+ * @param off_idx logical index of bitmask offset value into "prop" array
+ * @param sz_idx logical index of bitmask size value into "prop" array
+ */
```
Please separate the `@cond` statements from the rest of the comments as this allows for restructuring the functions/comments without accidentally moving the `@cond` to cover something else.

```suggestion
/**
 * @cond INTERNAL_HIDDEN
 *
 * For internal driver use only, skip these in public documentation.
 */

/** 
 * Auxiliary conditional macro that generates a bitmask for the range
 * from @p "prop" array defined by the (off_idx, sz_idx) pair,
 * or 0 if the range does not exist.
 *
 * @param node_id devicetree node identifier
 * @param prop lowercase-and-underscores array property name
 * @param off_idx logical index of bitmask offset value into "prop" array
 * @param sz_idx logical index of bitmask size value into "prop" array
 */
```

avolkov-1221 replied:
done

henrikbrixandersen commented on include/zephyr/drivers/gpio.h:
```diff
@@ -417,6 +417,247 @@ struct gpio_dt_spec {
+			odd_it)),					      \
+		(0))
+
+/**
+ * INTERNAL_HIDDEN @endcond
+ */
```
```suggestion
/**
 * @endcond
 */
```

avolkov-1221 replied:
done

henrikbrixandersen commented on include/zephyr/drivers/gpio.h:
```diff
@@ -417,6 +417,247 @@ struct gpio_dt_spec {
+ * @param ngpios  number of GPIOs
+ * @return the bitmask of allowed gpios
+ */
+#define GPIO_DT_PORT_PIN_MASK_NGPIOS_EXC(node_id, ngpios)		       \
```
Should the two `_EXC` macros be marked as internal? Is there a use case for these outside of calling `GPIO_PORT_PIN_MASK_FROM_DT_NODE()`?

avolkov-1221 replied:
If you will check my reverted modification of gpio drivers, this macro can be useful for the 'nonstandart' gpio drivers, like your gpio_xlnx_axi.c

henrikbrixandersen reviewed (changes requested):
A few minor requests/questions, otherwise looks good.
